### PR TITLE
Add basic HTTP Authorization

### LINKED ISSUE #32: Allow User/Password to be set on Exporter Requests
In order to secure metrics collection, add ability for User/Password to be set using basic auth during metric collection.

### PR DESCRIPTION
**Pull request by ColinSullivan1**
* Allows username/password to be set via command line.
* Bcrypt support
* Added tests

Resolves #32

### CODE REVIEW

**Comment by ColinSullivan1** [comment]
FYI - Closing and reopening to trigger a change with Travis.

File: exporter/exporter.go [hunk 1]
```diff
@@ -221,6 +229,58 @@ func (ne *NATSExporter) generateTLSConfig() (*tls.Config, error) {
...
+		return http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+			auth := strings.SplitN(r.Header.Get("Authorization"), " ", 2)
+			if len(auth) != 2 || auth[0] != "Basic" {
+				http.Error(rw, "authorization failed", http.StatusUnauthorized)
+				return
+			}
+
+			payload, err := base64.StdEncoding.DecodeString(auth[1])
+			if err != nil {
+				http.Error(rw, "authorization failed", http.StatusBadRequest)
```
Comment (wallyqs) [inline exporter/exporter.go]: missing `return` here I think?

Comment (ColinSullivan1) [reply exporter/exporter.go]: Good catch - Thanks!

File: exporter/exporter.go [hunk 2]
```diff
@@ -221,6 +229,59 @@ func (ne *NATSExporter) generateTLSConfig() (*tls.Config, error) {
...
+		if err := bcrypt.CompareHashAndPassword([]byte(exporterPassword), []byte(password)); err != nil {
+			return false
+		}
+	} else if exporterPassword != password {
+		return false
+	}
+	return true
+}
+
+// getScrapeHandler returns the defualt handler if no nttp
```
Comment (kozlovic) [inline exporter/exporter.go]: typo: `defualt`-> `default`

File: exporter/exporter_test.go [hunk 3]
```diff
@@ -421,3 +450,61 @@ func TestExporterStartNoMetricsSelected(t *testing.T) {
...
+
+	s := pet.RunServer()
+	defer s.Shutdown()
+
+	// first try user/pass with no auth.
+	testBasicAuth(t, opts, "colin", "password", http.StatusOK)
+
+	// now try user/pass
+	opts.HTTPUser = "colin"
+	opts.HTTPPassword = "password"
```
Comment (kozlovic) [inline exporter/exporter_test.go]: I don't understand this. The original `opts` has already set the user/password and we already tried `testBasicAuth()` 3 lines above.

File: exporter/exporter_test.go [hunk 4]
```diff
@@ -54,19 +54,44 @@ func getSecure(t *testing.T, url string) (*http.Response, error) {
...
 	} else {
-		resp, err = http.Get(fmt.Sprintf("http://%s/metrics", addr))
+		resp, err = http.Get(url)
 	}
 	if err != nil {
 		return err
 	}
-	if resp.StatusCode != 200 {
-		return fmt.Errorf("expected a 200 response, got %d", resp.StatusCode)
+	// Avoid EOF errors in Travis
```
Comment (kozlovic) [inline exporter/exporter_test.go]: Should have a `defer resp.Body.Close()` here instead.

File: exporter/exporter_test.go [hunk 5]
```diff
@@ -54,19 +54,44 @@ func getSecure(t *testing.T, url string) (*http.Response, error) {
...
+	if rc != expectedRc {
+		return fmt.Errorf("expected a %d response, got %d", expectedRc, rc)
+	}
+
+	// bail on auth error, etc.
+	if rc != 200 {
+		return nil
 	}
 
 	defer func() {
```
Comment (kozlovic) [inline exporter/exporter_test.go]: Would not need that here if you put the defer just after getting the `resp`.

File: exporter/exporter_test.go [hunk 6]
```diff
@@ -54,19 +54,44 @@ func getSecure(t *testing.T, url string) (*http.Response, error) {
...
+	resp.Close = true
+
+	rc := resp.StatusCode
+	if rc != expectedRc {
+		return fmt.Errorf("expected a %d response, got %d", expectedRc, rc)
+	}
+
+	// bail on auth error, etc.
+	if rc != 200 {
+		return nil
```
Comment (kozlovic) [inline exporter/exporter_test.go]: So we test above for the expectedRc, but if we are here and it is not a `200`, then we return no error. The close of `resp.Body` was missing. Having the `defer` above as mentioned will solve that.

File: exporter/exporter.go (hunk 2 above)
Comment (ColinSullivan1) [reply exporter/exporter.go]: Thanks!

File: exporter/exporter_test.go (hunk 3 above)
Comment (ColinSullivan1) [reply exporter/exporter_test.go]: It's a bug in the test...  thanks!

File: exporter/exporter_test.go (hunk 6 above)
Comment (ColinSullivan1) [reply exporter/exporter_test.go]: Ah, yep.  Thanks.

File: exporter/exporter_test.go [hunk 7]
```diff
@@ -421,3 +452,59 @@ func TestExporterStartNoMetricsSelected(t *testing.T) {
...
+		t.Fatalf("%v", err)
+	}
+	if err := testBasicAuth(t, opts, "", "password", http.StatusUnauthorized); err != nil {
+		t.Fatalf("%v", err)
+	}
+	if err := testBasicAuth(t, opts, "colin", "", http.StatusUnauthorized); err != nil {
+		t.Fatalf("%v", err)
+	}
+
+	opts.HTTPPassword = "$2a$11$SITqoMbjeKK1y9iV6nhXa.fC8o/QXEqs7o7DMkFtVErp0aTn12o1y"
```
Comment (kozlovic) [inline exporter/exporter_test.go]: Is that the cost 11? If so, way too high for Travis. You may get a timeout and cut off.

Comment (kozlovic) [reply exporter/exporter_test.go]: Try 5 max.

File: exporter/exporter_test.go [hunk 8]
```diff
@@ -46,33 +46,60 @@ func getSecure(t *testing.T, url string) (*http.Response, error) {
...
 	}
 	if err != nil {
-		return err
-	}
-	if resp.StatusCode != 200 {
-		return fmt.Errorf("expected a 200 response, got %d", resp.StatusCode)
+		return fmt.Errorf("error from get: %v", err)
 	}
 
 	defer func() {
```
Comment (kozlovic) [inline exporter/exporter_test.go]: Could be simplified with `defer resp.Body.Close()`

File: exporter/exporter.go (hunk 1 above)
Comment (aricart) [reply exporter/exporter.go]: I wouldn't do BadRequest on failing to decode the authorization. I would simply fail it with the Unauthorized

**Comment by kozlovic** [comment]
Also, you may need to call this: `http.DefaultTransport.(*http.Transport).CloseIdleConnections()` at beginning of each test or even at beginning of `testBasicAuth()` since this function is called many times in the same test.

**Comment by aricart** [comment]
A few things:

- I think adding another layer of credentials is probably inconsistent. If gnatsd monitoring doesn't provide it currently, we shouldn't cascade configuration that is only for this component.

- If we were cascading auth configuration from the gnatsd, then likely the authorization would need to come from gnatsd and transformed into some sort of JWT token that allows the request.

- I am not sure the point of bcrypt auth on non-tls connection to the server as the credentials would go readable over the network. Also in fast polling there would be significant load on the machine.

- Finally, the correct way to do the integration is to have some sort of monitoring hb from the server (or req/rep). This would centralize the sources of truth and server access.

**Comment by ColinSullivan1** [comment]
@aricart , thanks for the comments.  This is driven by the `basic_auth` feature of prometheus.  We could alternatively put nginx in front of everything, but this is convenient.

Regarding bcrypt, I'm rethinking it...  It'd be simpler to remove it and just use TLS.

**Comment by kozlovic** [comment]
Good, now only `TestExporterDefaultOptions` is failing because of IPv6, so that should be easy to fix.

**Comment by ColinSullivan1** [comment]
I'm going to remove bcrypt (after all this!).  It'll cause more problems (CPU wise) than the convenience it adds.  TLS should be sufficient.

**Comment by derekcollison** [comment]
bcrypt is so we do not store plaintext passwords, which TLS does not help against. We should keep it in.

**Comment by ColinSullivan1** [comment]
The only place we keep the password is in memory passed from the command line - one would need to see the command line or inspect a dump.  But it doesn't hurt to have bcrypt support so long as users keep the cost low.

**Comment by ColinSullivan1** [comment]
@derekcollison , bcrypt added back in.

**Comment by derekcollison** [comment]
We can use plaintext for tests too. That will avoid the high costs and timeouts.

**Comment by ColinSullivan1** [comment]
Good to merge?

**Comment by kozlovic** [comment]
LGTM

**Pull request merged**